File: app/src/main/java/eu/kanade/mangafeed/data/models/Page.java
package eu.kanade.mangafeed.data.models;

import eu.kanade.mangafeed.data.helpers.NetworkHelper;
import rx.subjects.BehaviorSubject;

public class Page implements NetworkHelper.ProgressListener {

    private int pageNumber;
    private String url;
    private String imageUrl;
    private String imagePath;
    private transient int status;
    private transient int progress;

    private transient BehaviorSubject<Integer> statusSubject;

    public static final int QUEUE = 0;
    public static final int LOAD_PAGE = 1;
    public static final int DOWNLOAD_IMAGE = 2;
    public static final int READY = 3;
    public static final int ERROR = 4;

    public Page(int pageNumber, String url, String imageUrl, String imagePath) {
        this.pageNumber = pageNumber;
        this.url = url;
        this.imageUrl = imageUrl;
        this.imagePath = imagePath;
    }

    public Page(int pageNumber, String url) {
        this(pageNumber, url, null, null);
    }

    public int getPageNumber() {
        return pageNumber;
    }

    public String getUrl() {
        return url;
    }

    public String getImageUrl() {
        return imageUrl;
    }

    public void setImageUrl(String imageUrl) {
        this.imageUrl = imageUrl;
    }

    public String getImagePath() {
        return imagePath;
    }

    public void setImagePath(String imagePath) {
        this.imagePath = imagePath;
    }

    public int getStatus() {
        return status;
    }

    public void setStatus(int status) {
        this.status = status;
        notifyStatus();
    }

    public int getProgress() {
        return progress;
    }

    @Override
    public void update(long bytesRead, long contentLength, boolean done) {
        progress = (int) ((100 * bytesRead) / contentLength);
    }

    public void setStatusSubject(BehaviorSubject<Integer> subject) {
        this.statusSubject = subject;
        notifyStatus();
    }

    private void notifyStatus() {
        if (statusSubject != null)
            statusSubject.onNext(status);
    }

}

File: app/src/main/java/eu/kanade/tachiyomi/source/model/Page.kt
package eu.kanade.tachiyomi.source.model

import android.net.Uri
import eu.kanade.tachiyomi.network.ProgressListener
import kotlinx.serialization.Serializable
import kotlinx.serialization.Transient
import rx.subjects.Subject
import tachiyomi.source.model.PageUrl

@Serializable
open class Page(
    val index: Int,
    val url: String = "",
    var imageUrl: String? = null,
    @Transient var uri: Uri? = null, // Deprecated but can't be deleted due to extensions
) : ProgressListener {

    val number: Int
        get() = index + 1

    @Transient
    @Volatile
    var status: Int = 0
        set(value) {
            field = value
            statusSubject?.onNext(value)
            statusCallback?.invoke(this)
        }

    @Transient
    @Volatile
    var progress: Int = 0
        set(value) {
            field = value
            statusCallback?.invoke(this)
        }

    @Transient
    private var statusSubject: Subject<Int, Int>? = null

    @Transient
    private var statusCallback: ((Page) -> Unit)? = null

    override fun update(bytesRead: Long, contentLength: Long, done: Boolean) {
        progress = if (contentLength > 0) {
            (100 * bytesRead / contentLength).toInt()
        } else {
            -1
        }
    }

    fun setStatusSubject(subject: Subject<Int, Int>?) {
        this.statusSubject = subject
    }

    fun setStatusCallback(f: ((Page) -> Unit)?) {
        statusCallback = f
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Page) return false

        if (index != other.index) return false
        if (url != other.url) return false
        if (imageUrl != other.imageUrl) return false
        if (number != other.number) return false
        if (status != other.status) return false
        if (progress != other.progress) return false
        if (statusSubject != other.statusSubject) return false
        if (statusCallback != other.statusCallback) return false

        return true
    }

    override fun hashCode(): Int {
        var result = index
        result = 31 * result + url.hashCode()
        result = 31 * result + (imageUrl?.hashCode() ?: 0)
        result = 31 * result + status
        result = 31 * result + progress
        result = 31 * result + (statusSubject?.hashCode() ?: 0)
        result = 31 * result + (statusCallback?.hashCode() ?: 0)
        return result
    }

    companion object {
        const val QUEUE = 0
        const val LOAD_PAGE = 1
        const val DOWNLOAD_IMAGE = 2
        const val READY = 3
        const val ERROR = 4
    }
}

fun Page.toPageUrl(): PageUrl {
    return PageUrl(
        url = this.imageUrl ?: this.url,
    )
}

fun PageUrl.toPage(index: Int): Page {
    return Page(
        index = index,
        imageUrl = this.url,
    )
}
